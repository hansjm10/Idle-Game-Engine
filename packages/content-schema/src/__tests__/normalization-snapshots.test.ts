/**
 * Snapshot tests verifying normalization determinism across multiple runs
 * per docs/content-dsl-schema-design.md §6 and §5.17.
 */

import { describe, expect, it } from 'vitest';
import { createContentPackValidator } from '../index.js';
import { validComprehensivePackFixture } from '../__fixtures__/integration-packs.js';

describe('Normalization Snapshots', () => {
  it('produces deterministic output for comprehensive pack', () => {
    const validator = createContentPackValidator({
      runtimeVersion: '1.0.0',
      activePackIds: ['@idle-engine/core'],
    });

    const result = validator.parse(validComprehensivePackFixture);

    // Snapshot the entire normalized pack
    expect(result.pack).toMatchSnapshot();

    // Verify digest is stable
    expect(result.pack.digest).toMatchSnapshot('pack-digest');
  });

  it('produces identical output across multiple parses of same input', () => {
    const validator = createContentPackValidator();

    const firstParse = validator.parse(validComprehensivePackFixture);
    const secondParse = validator.parse(
      JSON.parse(JSON.stringify(validComprehensivePackFixture)),
    );
    const thirdParse = validator.parse({
      ...validComprehensivePackFixture,
    });

    // All parses should produce identical digests
    expect(firstParse.pack.digest.hash).toBe(secondParse.pack.digest.hash);
    expect(secondParse.pack.digest.hash).toBe(thirdParse.pack.digest.hash);

    // Verify resources are in same order
    expect(firstParse.pack.resources.map((r) => r.id)).toEqual(
      secondParse.pack.resources.map((r) => r.id),
    );
  });

  it('normalizes resource ordering deterministically', () => {
    const validator = createContentPackValidator();

    const packWithUnorderedResources = {
      metadata: {
        id: 'order-test',
        title: { default: 'Order Test', variants: {} },
        version: '1.0.0',
        engine: '^1.0.0',
        defaultLocale: 'en-US',
        supportedLocales: ['en-US'],
      },
      resources: [
        {
          id: 'zulu',
          name: { default: 'Zulu', variants: {} },
          category: 'primary' as const,
          tier: 1,
          order: 3,
        },
        {
          id: 'alpha',
          name: { default: 'Alpha', variants: {} },
          category: 'primary' as const,
          tier: 1,
          order: 1,
        },
        {
          id: 'mike',
          name: { default: 'Mike', variants: {} },
          category: 'primary' as const,
          tier: 1,
          order: 2,
        },
      ],
      generators: [],
      upgrades: [],
    };

    const result = validator.parse(packWithUnorderedResources);

    // Verify order field determines sort
    expect(result.pack.resources.map((r) => r.id)).toEqual([
      'alpha',
      'mike',
      'zulu',
    ]);

    // Snapshot the ordering
    expect(result.pack.resources).toMatchSnapshot('ordered-resources');
  });

  it('normalizes locale variants deterministically', () => {
    const validator = createContentPackValidator();

    const packWithLocales = {
      metadata: {
        id: 'locale-test',
        title: {
          default: 'Locale Test',
          variants: {
            'fr-FR': 'Test de Locale',
            'es-ES': 'Prueba de Configuración Regional',
            'de-DE': 'Locale-Test',
          },
        },
        version: '1.0.0',
        engine: '^1.0.0',
        defaultLocale: 'en-US',
        supportedLocales: ['en-US', 'fr-FR', 'es-ES', 'de-DE'],
      },
      resources: [
        {
          id: 'energy',
          name: {
            default: 'Energy',
            variants: {
              'fr-FR': 'Énergie',
              'es-ES': 'Energía',
            },
          },
          category: 'primary' as const,
          tier: 1,
        },
      ],
      generators: [],
      upgrades: [],
    };

    const result = validator.parse(packWithLocales);

    // Snapshot normalized locales
    expect(result.pack.metadata.title).toMatchSnapshot('title-with-locales');
    expect(result.pack.resources[0]?.name).toMatchSnapshot(
      'resource-name-with-locales',
    );

    // Verify defaultLocale is mirrored into variants
    expect(result.pack.metadata.title.variants['en-US']).toBe('Locale Test');
  });

  it('normalizes formulas deterministically', () => {
    const validator = createContentPackValidator();

    const packWithFormulas = {
      metadata: {
        id: 'formula-test',
        title: { default: 'Formula Test', variants: {} },
        version: '1.0.0',
        engine: '^1.0.0',
        defaultLocale: 'en-US',
        supportedLocales: ['en-US'],
      },
      resources: [
        {
          id: 'energy',
          name: { default: 'Energy', variants: {} },
          category: 'primary' as const,
          tier: 1,
        },
      ],
      generators: [
        {
          id: 'solar-panel',
          name: { default: 'Solar Panel', variants: {} },
          produces: [
            {
              resourceId: 'energy',
              rate: {
                kind: 'piecewise' as const,
                pieces: [
                  {
                    untilLevel: 10,
                    formula: { kind: 'linear' as const, base: 1, slope: 0.5 },
                  },
                  {
                    untilLevel: 50,
                    formula: {
                      kind: 'exponential' as const,
                      base: 10,
                      growth: 1.1,
                      offset: 5,
                    },
                  },
                  {
                    formula: {
                      kind: 'polynomial' as const,
                      coefficients: [100, 10, 1],
                    },
                  },
                ],
              },
            },
          ],
          consumes: [],
          purchase: {
            currencyId: 'energy',
            baseCost: 10,
            costCurve: { kind: 'constant' as const, value: 10 },
          },
          baseUnlock: { kind: 'always' as const },
        },
      ],
      upgrades: [],
    };

    const result = validator.parse(packWithFormulas);

    // Snapshot the normalized formulas
    expect(result.pack.generators[0]?.produces[0]?.rate).toMatchSnapshot(
      'piecewise-formula',
    );
  });

  it('normalizes conditions deterministically', () => {
    const validator = createContentPackValidator();

    const packWithConditions = {
      metadata: {
        id: 'condition-test',
        title: { default: 'Condition Test', variants: {} },
        version: '1.0.0',
        engine: '^1.0.0',
        defaultLocale: 'en-US',
        supportedLocales: ['en-US'],
      },
      resources: [
        {
          id: 'energy',
          name: { default: 'Energy', variants: {} },
          category: 'primary' as const,
          tier: 1,
        },
        {
          id: 'crystals',
          name: { default: 'Crystals', variants: {} },
          category: 'prestige' as const,
          tier: 2,
          unlockCondition: {
            kind: 'allOf' as const,
            conditions: [
              {
                kind: 'resourceThreshold' as const,
                resourceId: 'energy',
                comparator: 'gte' as const,
                amount: { kind: 'constant' as const, value: 100 },
              },
              {
                kind: 'anyOf' as const,
                conditions: [
                  {
                    kind: 'resourceThreshold' as const,
                    resourceId: 'energy',
                    comparator: 'gte' as const,
                    amount: { kind: 'constant' as const, value: 200 },
                  },
                ],
              },
            ],
          },
        },
      ],
      generators: [],
      upgrades: [],
    };

    const result = validator.parse(packWithConditions);

    // Snapshot the normalized condition tree
    expect(
      result.pack.resources.find((r) => r.id === 'crystals')?.unlockCondition,
    ).toMatchSnapshot('complex-condition-tree');
  });

  it('normalizes lookup maps structure', () => {
    const validator = createContentPackValidator();

    const result = validator.parse(validComprehensivePackFixture);

    // Snapshot lookup map keys
    const lookupKeys = {
      resources: Array.from(result.pack.lookup.resources.keys()),
      generators: Array.from(result.pack.lookup.generators.keys()),
      upgrades: Array.from(result.pack.lookup.upgrades.keys()),
      metrics: Array.from(result.pack.lookup.metrics.keys()),
      achievements: Array.from(result.pack.lookup.achievements.keys()),
    };

    expect(lookupKeys).toMatchSnapshot('lookup-map-keys');

    // Snapshot serialized lookup keys
    const serializedLookupKeys = {
      resources: Object.keys(result.pack.serializedLookup.resourceById),
      generators: Object.keys(result.pack.serializedLookup.generatorById),
      upgrades: Object.keys(result.pack.serializedLookup.upgradeById),
    };

    expect(serializedLookupKeys).toMatchSnapshot('serialized-lookup-keys');
  });

  it('normalizes metadata fields deterministically', () => {
    const validator = createContentPackValidator();

    const packWithMetadata = {
      metadata: {
        id: 'metadata-test',
        title: { default: 'Metadata Test', variants: {} },
        version: '1.0.0',
        engine: '^1.0.0',
        defaultLocale: 'en-US',
        supportedLocales: ['fr-FR', 'en-US', 'es-ES'], // Unsorted
        authors: ['Author B', 'Author A', 'Author C'], // Unsorted
        tags: ['zulu', 'alpha', 'mike'], // Unsorted
      },
      resources: [],
      generators: [],
      upgrades: [],
    };

    const result = validator.parse(packWithMetadata);

    // Snapshot normalized metadata (should be sorted)
    expect(result.pack.metadata).toMatchSnapshot('normalized-metadata');

    // Verify arrays are sorted
    expect(result.pack.metadata.supportedLocales).toEqual([
      'en-US',
      'es-ES',
      'fr-FR',
    ]);
  });

  it('produces consistent digest hashes for equivalent packs', () => {
    const validator = createContentPackValidator();

    const pack1 = {
      metadata: {
        id: 'test-pack',
        title: { default: 'Test', variants: {} },
        version: '1.0.0',
        engine: '^1.0.0',
        defaultLocale: 'en-US',
        supportedLocales: ['en-US'],
      },
      resources: [
        {
          id: 'energy',
          name: { default: 'Energy', variants: {} },
          category: 'primary' as const,
          tier: 1,
        },
      ],
      generators: [],
      upgrades: [],
    };

    // Create equivalent pack with different ordering
    const pack2 = JSON.parse(JSON.stringify(pack1));

    const result1 = validator.parse(pack1);
    const result2 = validator.parse(pack2);

    // Digests should match
    expect(result1.pack.digest.hash).toBe(result2.pack.digest.hash);
    expect(result1.pack.digest.version).toBe(result2.pack.digest.version);

    // Snapshot the digest format
    expect(result1.pack.digest).toMatchSnapshot('digest-format');
  });
});
