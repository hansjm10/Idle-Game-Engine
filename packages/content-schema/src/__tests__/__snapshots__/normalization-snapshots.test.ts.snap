// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Normalization Snapshots > normalizes conditions deterministically > complex-condition-tree 1`] = `
{
  "conditions": [
    {
      "amount": {
        "kind": "constant",
        "value": 100,
      },
      "comparator": "gte",
      "kind": "resourceThreshold",
      "resourceId": "energy",
    },
    {
      "conditions": [
        {
          "amount": {
            "kind": "constant",
            "value": 200,
          },
          "comparator": "gte",
          "kind": "resourceThreshold",
          "resourceId": "energy",
        },
      ],
      "kind": "anyOf",
    },
  ],
  "kind": "allOf",
}
`;

exports[`Normalization Snapshots > normalizes formulas deterministically > piecewise-formula 1`] = `
{
  "kind": "piecewise",
  "pieces": [
    {
      "formula": {
        "base": 1,
        "kind": "linear",
        "slope": 0.5,
      },
      "untilLevel": 10,
    },
    {
      "formula": {
        "base": 10,
        "growth": 1.1,
        "kind": "exponential",
        "offset": 5,
      },
      "untilLevel": 50,
    },
    {
      "formula": {
        "coefficients": [
          100,
          10,
          1,
        ],
        "kind": "polynomial",
      },
    },
  ],
}
`;

exports[`Normalization Snapshots > normalizes locale variants deterministically > resource-name-with-locales 1`] = `
{
  "default": "Energy",
  "variants": {
    "en-US": "Energy",
    "es-ES": "Energía",
    "fr-FR": "Énergie",
  },
}
`;

exports[`Normalization Snapshots > normalizes locale variants deterministically > title-with-locales 1`] = `
{
  "default": "Locale Test",
  "variants": {
    "de-DE": "Locale-Test",
    "en-US": "Locale Test",
    "es-ES": "Prueba de Configuración Regional",
    "fr-FR": "Test de Locale",
  },
}
`;

exports[`Normalization Snapshots > normalizes lookup maps structure > lookup-map-keys 1`] = `
{
  "achievements": [
    "first-energy",
  ],
  "generators": [
    "solar-panel",
  ],
  "metrics": [
    "total-energy-produced",
  ],
  "resources": [
    "energy",
    "crystals",
  ],
  "upgrades": [
    "efficiency-boost",
  ],
}
`;

exports[`Normalization Snapshots > normalizes lookup maps structure > serialized-lookup-keys 1`] = `
{
  "generators": [
    "solar-panel",
  ],
  "resources": [
    "energy",
    "crystals",
  ],
  "upgrades": [
    "efficiency-boost",
  ],
}
`;

exports[`Normalization Snapshots > normalizes metadata fields deterministically > normalized-metadata 1`] = `
{
  "authors": [
    "Author A",
    "Author B",
    "Author C",
  ],
  "defaultLocale": "en-US",
  "engine": ">=1.0.0 <2.0.0-0",
  "id": "metadata-test",
  "links": [],
  "summary": undefined,
  "supportedLocales": [
    "en-US",
    "es-ES",
    "fr-FR",
  ],
  "tags": [
    "alpha",
    "mike",
    "zulu",
  ],
  "title": {
    "default": "Metadata Test",
    "variants": {
      "en-US": "Metadata Test",
    },
  },
  "version": "1.0.0",
}
`;

exports[`Normalization Snapshots > normalizes resource ordering deterministically > ordered-resources 1`] = `
[
  {
    "capacity": 0,
    "category": "primary",
    "id": "alpha",
    "name": {
      "default": "Alpha",
      "variants": {
        "en-US": "Alpha",
      },
    },
    "order": 1,
    "startAmount": 0,
    "tags": [],
    "tier": 1,
    "unlocked": false,
    "visible": true,
  },
  {
    "capacity": 0,
    "category": "primary",
    "id": "mike",
    "name": {
      "default": "Mike",
      "variants": {
        "en-US": "Mike",
      },
    },
    "order": 2,
    "startAmount": 0,
    "tags": [],
    "tier": 1,
    "unlocked": false,
    "visible": true,
  },
  {
    "capacity": 0,
    "category": "primary",
    "id": "zulu",
    "name": {
      "default": "Zulu",
      "variants": {
        "en-US": "Zulu",
      },
    },
    "order": 3,
    "startAmount": 0,
    "tags": [],
    "tier": 1,
    "unlocked": false,
    "visible": true,
  },
]
`;

exports[`Normalization Snapshots > produces consistent digest hashes for equivalent packs > digest-format 1`] = `
{
  "hash": "fnv1a-0869eed8",
  "version": 1,
}
`;

exports[`Normalization Snapshots > produces deterministic output for comprehensive pack > pack-digest 1`] = `
{
  "hash": "fnv1a-d11b94ec",
  "version": 1,
}
`;

exports[`Normalization Snapshots > produces deterministic output for comprehensive pack 1`] = `
{
  "achievements": [
    {
      "category": "progression",
      "description": {
        "default": "Produce your first unit of energy",
        "variants": {
          "en-US": "Produce your first unit of energy",
          "fr-FR": "Produisez votre première unité d'énergie",
        },
      },
      "id": "first-energy",
      "name": {
        "default": "First Energy",
        "variants": {
          "en-US": "First Energy",
          "fr-FR": "Première Énergie",
        },
      },
      "onUnlockEvents": [],
      "progress": {
        "mode": "oneShot",
        "repeatable": undefined,
        "target": {
          "kind": "constant",
          "value": 1,
        },
      },
      "tags": [],
      "tier": "bronze",
      "track": {
        "comparator": "gte",
        "kind": "resource",
        "resourceId": "energy",
        "threshold": {
          "kind": "constant",
          "value": 1,
        },
      },
    },
  ],
  "automations": [],
  "digest": {
    "hash": "fnv1a-d11b94ec",
    "version": 1,
  },
  "generators": [
    {
      "baseUnlock": {
        "kind": "always",
      },
      "consumes": [],
      "effects": [],
      "id": "solar-panel",
      "name": {
        "default": "Solar Panel",
        "variants": {
          "en-US": "Solar Panel",
          "fr-FR": "Panneau Solaire",
        },
      },
      "order": 1,
      "produces": [
        {
          "rate": {
            "base": 1,
            "kind": "linear",
            "slope": 0.5,
          },
          "resourceId": "energy",
        },
      ],
      "purchase": {
        "baseCost": 10,
        "costCurve": {
          "base": 10,
          "growth": 1.15,
          "kind": "exponential",
          "offset": 0,
        },
        "currencyId": "energy",
      },
      "tags": [],
    },
  ],
  "guildPerks": [],
  "lookup": {
    "achievements": Map {
      "first-energy" => {
        "category": "progression",
        "description": {
          "default": "Produce your first unit of energy",
          "variants": {
            "en-US": "Produce your first unit of energy",
            "fr-FR": "Produisez votre première unité d'énergie",
          },
        },
        "id": "first-energy",
        "name": {
          "default": "First Energy",
          "variants": {
            "en-US": "First Energy",
            "fr-FR": "Première Énergie",
          },
        },
        "onUnlockEvents": [],
        "progress": {
          "mode": "oneShot",
          "repeatable": undefined,
          "target": {
            "kind": "constant",
            "value": 1,
          },
        },
        "tags": [],
        "tier": "bronze",
        "track": {
          "comparator": "gte",
          "kind": "resource",
          "resourceId": "energy",
          "threshold": {
            "kind": "constant",
            "value": 1,
          },
        },
      },
    },
    "automations": Map {},
    "generators": Map {
      "solar-panel" => {
        "baseUnlock": {
          "kind": "always",
        },
        "consumes": [],
        "effects": [],
        "id": "solar-panel",
        "name": {
          "default": "Solar Panel",
          "variants": {
            "en-US": "Solar Panel",
            "fr-FR": "Panneau Solaire",
          },
        },
        "order": 1,
        "produces": [
          {
            "rate": {
              "base": 1,
              "kind": "linear",
              "slope": 0.5,
            },
            "resourceId": "energy",
          },
        ],
        "purchase": {
          "baseCost": 10,
          "costCurve": {
            "base": 10,
            "growth": 1.15,
            "kind": "exponential",
            "offset": 0,
          },
          "currencyId": "energy",
        },
        "tags": [],
      },
    },
    "guildPerks": Map {},
    "metrics": Map {
      "total-energy-produced" => {
        "attributes": [],
        "description": {
          "default": "Cumulative energy produced across all time",
          "variants": {
            "en-US": "Cumulative energy produced across all time",
          },
        },
        "id": "total-energy-produced",
        "kind": "counter",
        "name": {
          "default": "Total Energy Produced",
          "variants": {
            "en-US": "Total Energy Produced",
            "fr-FR": "Énergie Totale Produite",
          },
        },
        "source": {
          "kind": "runtime",
        },
        "unit": "units",
      },
    },
    "prestigeLayers": Map {},
    "resources": Map {
      "energy" => {
        "capacity": 1000,
        "category": "primary",
        "id": "energy",
        "name": {
          "default": "Energy",
          "variants": {
            "en-US": "Energy",
            "es-ES": "Energía",
            "fr-FR": "Énergie",
          },
        },
        "order": 1,
        "startAmount": 10,
        "tags": [],
        "tier": 1,
        "unlocked": true,
        "visible": true,
      },
      "crystals" => {
        "capacity": 0,
        "category": "prestige",
        "id": "crystals",
        "name": {
          "default": "Crystals",
          "variants": {
            "en-US": "Crystals",
            "es-ES": "Cristales",
            "fr-FR": "Cristaux",
          },
        },
        "order": 2,
        "startAmount": 0,
        "tags": [],
        "tier": 2,
        "unlockCondition": {
          "amount": {
            "kind": "constant",
            "value": 100,
          },
          "comparator": "gte",
          "kind": "resourceThreshold",
          "resourceId": "energy",
        },
        "unlocked": false,
        "visible": true,
      },
    },
    "runtimeEvents": Map {},
    "transforms": Map {},
    "upgrades": Map {
      "efficiency-boost" => {
        "category": "global",
        "cost": {
          "baseCost": 50,
          "costCurve": {
            "kind": "constant",
            "value": 50,
          },
          "currencyId": "energy",
        },
        "effects": [
          {
            "generatorId": "solar-panel",
            "kind": "modifyGeneratorRate",
            "operation": "multiply",
            "value": {
              "kind": "constant",
              "value": 1.5,
            },
          },
        ],
        "id": "efficiency-boost",
        "name": {
          "default": "Efficiency Boost",
          "variants": {
            "en-US": "Efficiency Boost",
            "fr-FR": "Amélioration d'Efficacité",
          },
        },
        "prerequisites": [],
        "tags": [],
        "targets": [
          {
            "kind": "global",
          },
        ],
        "unlockCondition": {
          "amount": {
            "kind": "constant",
            "value": 50,
          },
          "comparator": "gte",
          "kind": "resourceThreshold",
          "resourceId": "energy",
        },
      },
    },
  },
  "metadata": {
    "authors": [
      "Another Author",
      "Test Author",
    ],
    "defaultLocale": "en-US",
    "dependencies": {
      "conflicts": [],
      "optional": [
        {
          "packId": "optional-pack",
          "version": ">=1.0.0 <2.0.0-0",
        },
      ],
      "provides": [],
      "requires": [
        {
          "packId": "@idle-engine/core",
          "version": ">=1.0.0 <2.0.0-0",
        },
      ],
    },
    "engine": ">=1.0.0 <2.0.0-0",
    "id": "comprehensive-test",
    "links": [],
    "summary": {
      "default": "A complete pack covering all modules",
      "variants": {
        "en-US": "A complete pack covering all modules",
        "fr-FR": "Un pack complet couvrant tous les modules",
      },
    },
    "supportedLocales": [
      "en-US",
      "es-ES",
      "fr-FR",
    ],
    "tags": [
      "comprehensive",
      "test",
    ],
    "title": {
      "default": "Comprehensive Test Pack",
      "variants": {
        "en-US": "Comprehensive Test Pack",
        "es-ES": "Paquete de Prueba Integral",
        "fr-FR": "Pack de Test Complet",
      },
    },
    "version": "1.0.0",
  },
  "metrics": [
    {
      "attributes": [],
      "description": {
        "default": "Cumulative energy produced across all time",
        "variants": {
          "en-US": "Cumulative energy produced across all time",
        },
      },
      "id": "total-energy-produced",
      "kind": "counter",
      "name": {
        "default": "Total Energy Produced",
        "variants": {
          "en-US": "Total Energy Produced",
          "fr-FR": "Énergie Totale Produite",
        },
      },
      "source": {
        "kind": "runtime",
      },
      "unit": "units",
    },
  ],
  "prestigeLayers": [],
  "resources": [
    {
      "capacity": 1000,
      "category": "primary",
      "id": "energy",
      "name": {
        "default": "Energy",
        "variants": {
          "en-US": "Energy",
          "es-ES": "Energía",
          "fr-FR": "Énergie",
        },
      },
      "order": 1,
      "startAmount": 10,
      "tags": [],
      "tier": 1,
      "unlocked": true,
      "visible": true,
    },
    {
      "capacity": 0,
      "category": "prestige",
      "id": "crystals",
      "name": {
        "default": "Crystals",
        "variants": {
          "en-US": "Crystals",
          "es-ES": "Cristales",
          "fr-FR": "Cristaux",
        },
      },
      "order": 2,
      "startAmount": 0,
      "tags": [],
      "tier": 2,
      "unlockCondition": {
        "amount": {
          "kind": "constant",
          "value": 100,
        },
        "comparator": "gte",
        "kind": "resourceThreshold",
        "resourceId": "energy",
      },
      "unlocked": false,
      "visible": true,
    },
  ],
  "runtimeEvents": [],
  "serializedLookup": {
    "achievementById": {
      "first-energy": {
        "category": "progression",
        "description": {
          "default": "Produce your first unit of energy",
          "variants": {
            "en-US": "Produce your first unit of energy",
            "fr-FR": "Produisez votre première unité d'énergie",
          },
        },
        "id": "first-energy",
        "name": {
          "default": "First Energy",
          "variants": {
            "en-US": "First Energy",
            "fr-FR": "Première Énergie",
          },
        },
        "onUnlockEvents": [],
        "progress": {
          "mode": "oneShot",
          "repeatable": undefined,
          "target": {
            "kind": "constant",
            "value": 1,
          },
        },
        "tags": [],
        "tier": "bronze",
        "track": {
          "comparator": "gte",
          "kind": "resource",
          "resourceId": "energy",
          "threshold": {
            "kind": "constant",
            "value": 1,
          },
        },
      },
    },
    "automationById": {},
    "generatorById": {
      "solar-panel": {
        "baseUnlock": {
          "kind": "always",
        },
        "consumes": [],
        "effects": [],
        "id": "solar-panel",
        "name": {
          "default": "Solar Panel",
          "variants": {
            "en-US": "Solar Panel",
            "fr-FR": "Panneau Solaire",
          },
        },
        "order": 1,
        "produces": [
          {
            "rate": {
              "base": 1,
              "kind": "linear",
              "slope": 0.5,
            },
            "resourceId": "energy",
          },
        ],
        "purchase": {
          "baseCost": 10,
          "costCurve": {
            "base": 10,
            "growth": 1.15,
            "kind": "exponential",
            "offset": 0,
          },
          "currencyId": "energy",
        },
        "tags": [],
      },
    },
    "guildPerkById": {},
    "metricById": {
      "total-energy-produced": {
        "attributes": [],
        "description": {
          "default": "Cumulative energy produced across all time",
          "variants": {
            "en-US": "Cumulative energy produced across all time",
          },
        },
        "id": "total-energy-produced",
        "kind": "counter",
        "name": {
          "default": "Total Energy Produced",
          "variants": {
            "en-US": "Total Energy Produced",
            "fr-FR": "Énergie Totale Produite",
          },
        },
        "source": {
          "kind": "runtime",
        },
        "unit": "units",
      },
    },
    "prestigeLayerById": {},
    "resourceById": {
      "crystals": {
        "capacity": 0,
        "category": "prestige",
        "id": "crystals",
        "name": {
          "default": "Crystals",
          "variants": {
            "en-US": "Crystals",
            "es-ES": "Cristales",
            "fr-FR": "Cristaux",
          },
        },
        "order": 2,
        "startAmount": 0,
        "tags": [],
        "tier": 2,
        "unlockCondition": {
          "amount": {
            "kind": "constant",
            "value": 100,
          },
          "comparator": "gte",
          "kind": "resourceThreshold",
          "resourceId": "energy",
        },
        "unlocked": false,
        "visible": true,
      },
      "energy": {
        "capacity": 1000,
        "category": "primary",
        "id": "energy",
        "name": {
          "default": "Energy",
          "variants": {
            "en-US": "Energy",
            "es-ES": "Energía",
            "fr-FR": "Énergie",
          },
        },
        "order": 1,
        "startAmount": 10,
        "tags": [],
        "tier": 1,
        "unlocked": true,
        "visible": true,
      },
    },
    "runtimeEventById": {},
    "transformById": {},
    "upgradeById": {
      "efficiency-boost": {
        "category": "global",
        "cost": {
          "baseCost": 50,
          "costCurve": {
            "kind": "constant",
            "value": 50,
          },
          "currencyId": "energy",
        },
        "effects": [
          {
            "generatorId": "solar-panel",
            "kind": "modifyGeneratorRate",
            "operation": "multiply",
            "value": {
              "kind": "constant",
              "value": 1.5,
            },
          },
        ],
        "id": "efficiency-boost",
        "name": {
          "default": "Efficiency Boost",
          "variants": {
            "en-US": "Efficiency Boost",
            "fr-FR": "Amélioration d'Efficacité",
          },
        },
        "prerequisites": [],
        "tags": [],
        "targets": [
          {
            "kind": "global",
          },
        ],
        "unlockCondition": {
          "amount": {
            "kind": "constant",
            "value": 50,
          },
          "comparator": "gte",
          "kind": "resourceThreshold",
          "resourceId": "energy",
        },
      },
    },
  },
  "transforms": [],
  "upgrades": [
    {
      "category": "global",
      "cost": {
        "baseCost": 50,
        "costCurve": {
          "kind": "constant",
          "value": 50,
        },
        "currencyId": "energy",
      },
      "effects": [
        {
          "generatorId": "solar-panel",
          "kind": "modifyGeneratorRate",
          "operation": "multiply",
          "value": {
            "kind": "constant",
            "value": 1.5,
          },
        },
      ],
      "id": "efficiency-boost",
      "name": {
        "default": "Efficiency Boost",
        "variants": {
          "en-US": "Efficiency Boost",
          "fr-FR": "Amélioration d'Efficacité",
        },
      },
      "prerequisites": [],
      "tags": [],
      "targets": [
        {
          "kind": "global",
        },
      ],
      "unlockCondition": {
        "amount": {
          "kind": "constant",
          "value": 50,
        },
        "comparator": "gte",
        "kind": "resourceThreshold",
        "resourceId": "energy",
      },
    },
  ],
}
`;
